File: examples/backend-node/src/webhooks.ts
import { createHash } from "node:crypto";

import type { VoidhashWebhookEvent } from "@voidhash/node";

import type { EntitlementsCache } from "./entitlements-cache";
import type { NoteStore } from "./notes";

const MAX_TRACKED_DELIVERIES = 1_000;

export type WebhookProcessor = {
  /** Applies a verified delivery. Safe to call twice with the same delivery. */
  readonly process: (event: VoidhashWebhookEvent, rawBody: string) => Promise<void>;
};

const stringField = (payload: unknown, field: string): string | undefined => {
  const value = (payload as Record<string, unknown> | null)?.[field];

  return typeof value === "string" ? value : undefined;
};

/**
 * A key that is identical across retries of the same delivery.
 *
 * Voidhash sends the bare payload with no envelope and no delivery id, so the
 * key comes from the payload's own subject (`subscriptionId` / `purchaseId` /
 * `personId`) plus `occurredAt`. When a payload carries none of those — a
 * `test.ping`, or an event type newer than this code — the raw body is hashed
 * instead: a retry re-signs with a fresh timestamp but re-sends the same bytes.
 */
const deliveryKey = (event: VoidhashWebhookEvent, rawBody: string): string => {
  const subject =
    stringField(event.payload, "subscriptionId") ??
    stringField(event.payload, "purchaseId") ??
    stringField(event.payload, "personId");

  if (subject === undefined) {
    return `${event.type}:sha256:${createHash("sha256").update(rawBody).digest("hex")}`;
  }

  return `${event.type}:${subject}:${stringField(event.payload, "occurredAt") ?? ""}`;
};

export type WebhookProcessorOptions = {
  readonly entitlements: EntitlementsCache;
  readonly notes: NoteStore;
};

/**
 * Handles verified webhook deliveries idempotently.
 *
 * Anything slower than 30s (or outside 2xx) is retried four times, so the same
 * delivery does arrive twice in practice. The seen-set below is the whole
 * defence; in a real service it belongs in Redis or a unique index, not in
 * process memory.
 */
export const createWebhookProcessor = (options: WebhookProcessorOptions): WebhookProcessor => {
  const { entitlements, notes } = options;

  // A Set iterates in insertion order, which makes eviction of the oldest key a
  // one-liner and keeps the set from growing without bound.
  const seen = new Set<string>();

  const remember = (key: string): void => {
    seen.add(key);

    if (seen.size > MAX_TRACKED_DELIVERIES) {
      const oldest = seen.values().next().value;

      if (oldest !== undefined) {
        seen.delete(oldest);
      }
    }
  };

  return {
    process: async (event, rawBody) => {
      const key = deliveryKey(event, rawBody);

      if (seen.has(key)) {
        console.log(`[webhook] ${event.type} ${key} already handled — ignoring the retry.`);

        return;
      }

      remember(key);

      const distinctId = stringField(event.payload, "distinctId");

      switch (event.type) {
        case "subscription.created":
        case "subscription.renewed":
        case "subscription.cancelled":
        case "subscription.expired":
        case "purchase.completed":
        case "purchase.refunded": {
          // Access just changed, so the cached answer is wrong. Dropping it is
          // enough — the next request reads through and repopulates it.
          if (distinctId !== undefined) {
            entitlements.invalidate(distinctId);
          }

          console.log(`[webhook] ${event.type} for ${distinctId ?? "unknown person"}.`);
          break;
        }

        case "person.deleted": {
          if (distinctId !== undefined) {
            entitlements.invalidate(distinctId);
            notes.forget(distinctId);
          }

          console.log(`[webhook] ${event.type} — dropped local data for ${distinctId ?? "?"}.`);
          break;
        }

        case "person.created":
        case "person.updated":
        case "test.ping": {
          console.log(`[webhook] ${event.type} received.`);
          break;
        }

        // A newer server can send event names this code has never heard of.
        // Acknowledge them rather than failing the delivery.
        default: {
          console.log(`[webhook] unhandled event type "${event.type}" — acknowledged.`);
        }
      }
    },
  };
};

File: examples/backend-go/webhooks.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"

	voidhash "github.com/voidhashcom/voidhash-go"
)

const (
	// maxWebhookBody caps the body a delivery may carry. Voidhash payloads are
	// a few hundred bytes.
	maxWebhookBody = 1 << 20
	// dedupeWindow is how long a handled delivery is remembered. It comfortably
	// covers the first two entries of the Voidhash retry schedule (1min, 5min).
	dedupeWindow = 15 * time.Minute
)

// webhookAck is the body returned to Voidhash. Anything 2xx marks the delivery
// as succeeded; the fields are for whoever reads the delivery log in Studio.
type webhookAck struct {
	Received  bool   `json:"received"`
	Event     string `json:"event"`
	Duplicate bool   `json:"duplicate,omitempty"`
}

// webhookSubject is the slice of a lifecycle payload this app cares about.
// Every purchase, subscription and person event carries the distinct id of the
// person it concerns; the rest of the payload differs per event type.
type webhookSubject struct {
	Type        string  `json:"type"`
	DistinctID  string  `json:"distinctId"`
	PersonID    string  `json:"personId"`
	ProductSlug *string `json:"productSlug"`
	OccurredAt  string  `json:"occurredAt"`
}

func (s *server) handleWebhook(w http.ResponseWriter, r *http.Request) {
	if s.webhookSecret == "" {
		writeError(w, http.StatusServiceUnavailable, "webhooks_not_configured",
			"set VOIDHASH_WEBHOOK_SECRET to the endpoint signing secret from Studio")
		return
	}

	// The signature covers the exact bytes Voidhash sent, so the body is read
	// once and passed through verbatim — never re-encoded from a decoded value.
	body, err := io.ReadAll(http.MaxBytesReader(w, r.Body, maxWebhookBody))
	if err != nil {
		writeError(w, http.StatusRequestEntityTooLarge, "payload_too_large",
			fmt.Sprintf("webhook bodies are capped at %d bytes", maxWebhookBody))
		return
	}

	event, err := voidhash.ConstructWebhookEvent(body, r.Header, s.webhookSecret)
	if err != nil {
		var verificationErr *voidhash.WebhookVerificationError
		if errors.As(err, &verificationErr) {
			// Retrying never fixes a bad signature, so answer 4xx and let
			// Voidhash mark the delivery failed instead of backing off.
			s.logger.Warn("rejected webhook", "reason", verificationErr.Reason, "detail", verificationErr.Detail)
			writeError(w, http.StatusBadRequest, "invalid_webhook", verificationErr.Reason)
			return
		}
		writeError(w, http.StatusBadRequest, "invalid_webhook", err.Error())
		return
	}

	key := deliveryKey(event.Type, body)
	if !s.deliveries.Claim(key) {
		s.logger.Info("ignoring duplicate webhook delivery", "event", event.Type)
		writeJSON(w, http.StatusOK, webhookAck{Received: true, Event: event.Type, Duplicate: true})
		return
	}

	// Acknowledge before doing the work: Voidhash gives a delivery 30 seconds
	// and retries anything slower, which is exactly how one purchase turns into
	// two handled events. The dedupe claim above is already taken, so a retry
	// that overlaps this handler is recognised as a duplicate.
	writeJSON(w, http.StatusOK, webhookAck{Received: true, Event: event.Type})
	go s.dispatchWebhook(event, key)
}

func (s *server) dispatchWebhook(event *voidhash.WebhookEvent, key string) {
	if err := s.applyWebhook(event); err != nil {
		// Give the claim back so the next Voidhash retry is processed instead
		// of being swallowed as a duplicate of a delivery that never landed.
		s.deliveries.Release(key)
		s.logger.Error("webhook handling failed", "event", event.Type, "error", err)
		return
	}
	s.logger.Info("webhook handled", "event", event.Type)
}

func (s *server) applyWebhook(event *voidhash.WebhookEvent) error {
	switch event.Type {
	case voidhash.WebhookTestPing:
		s.logger.Info("webhook test ping received", "timestamp", event.Timestamp)
		return nil

	case voidhash.WebhookEventPurchaseCompleted,
		voidhash.WebhookEventPurchaseRefunded,
		voidhash.WebhookEventSubscriptionCreated,
		voidhash.WebhookEventSubscriptionRenewed,
		voidhash.WebhookEventSubscriptionCancelled,
		voidhash.WebhookEventSubscriptionExpired,
		voidhash.WebhookEventPersonCreated,
		voidhash.WebhookEventPersonUpdated,
		voidhash.WebhookEventPersonDeleted:

		var subject webhookSubject
		if err := json.Unmarshal(event.Payload, &subject); err != nil {
			return fmt.Errorf("decoding %s payload: %w", event.Type, err)
		}
		if subject.DistinctID == "" {
			return fmt.Errorf("%s payload carries no distinctId", event.Type)
		}

		// Entitlements just changed upstream, so the cached answer is wrong
		// right now rather than in 60 seconds.
		s.entitlements.Invalidate(subject.DistinctID)
		if event.Type == voidhash.WebhookEventPersonDeleted {
			s.notes.Forget(subject.DistinctID)
		}
		s.logger.Info("entitlements invalidated",
			"event", event.Type,
			"distinctId", subject.DistinctID,
			"productSlug", derefOr(subject.ProductSlug, ""),
		)
		return nil

	default:
		// Voidhash can add event types without asking; unknown ones are
		// acknowledged and ignored rather than retried forever.
		s.logger.Info("ignoring unhandled webhook event", "event", event.Type)
		return nil
	}
}

// deliveryKey derives a retry-stable identity for a delivery.
//
// Voidhash sends no delivery id header, and each retry re-signs with a fresh
// timestamp, so neither the signature nor the timestamp can be used. The
// payload bytes are identical across attempts of the same delivery, so their
// digest — scoped by event type — is the stable key.
func deliveryKey(eventType string, body []byte) string {
	digest := sha256.New()
	digest.Write([]byte(eventType))
	digest.Write([]byte("."))
	digest.Write(body)
	return hex.EncodeToString(digest.Sum(nil))
}

// dedupeSet remembers recently handled deliveries.
//
// It rotates two generations rather than expiring keys individually: a key
// lives for somewhere between one and two windows, memory stays bounded by the
// delivery rate, and neither Claim nor Release ever scans the map.
type dedupeSet struct {
	window time.Duration

	mu        sync.Mutex
	rotatedAt time.Time
	current   map[string]struct{}
	previous  map[string]struct{}
}

func newDedupeSet(window time.Duration) *dedupeSet {
	return &dedupeSet{
		window:    window,
		rotatedAt: time.Now(),
		current:   map[string]struct{}{},
		previous:  map[string]struct{}{},
	}
}

// Claim records a delivery and reports whether the caller now owns it. A
// second call with the same key returns false until the key ages out.
func (d *dedupeSet) Claim(key string) bool {
	d.mu.Lock()
	defer d.mu.Unlock()

	d.rotateLocked(time.Now())
	if _, seen := d.current[key]; seen {
		return false
	}
	if _, seen := d.previous[key]; seen {
		return false
	}
	d.current[key] = struct{}{}
	return true
}

// Release forgets a delivery, so a later retry of it is handled afresh.
func (d *dedupeSet) Release(key string) {
	d.mu.Lock()
	defer d.mu.Unlock()

	delete(d.current, key)
	delete(d.previous, key)
}

func (d *dedupeSet) rotateLocked(now time.Time) {
	if now.Sub(d.rotatedAt) < d.window {
		return
	}
	d.previous = d.current
	d.current = map[string]struct{}{}
	d.rotatedAt = now
}

func derefOr[T any](value *T, fallback T) T {
	if value == nil {
		return fallback
	}
	return *value
}
